### PR TITLE
Add HTTPS upgrade checks

### PR DESCRIPTION
**Pull request by cuba**
Resolves https://github.com/brave/brave-browser/issues/36408

Security review: https://github.com/brave/reviews/issues/1593

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Check that http://example.com upgrades correctly. Back and forth navigation buttons work as expected
2. Check that some made up website (ex: http://jfdskllfsjlks.com) does not upgrade. Back and forth buttons work as expected.

Note: I don't know of any sites that do exist in http but not in https. If you are able to come across one please test it.

### CODE REVIEW

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift [hunk 1]
```diff
@@ -599,6 +599,26 @@ extension BrowserViewController: WKNavigationDelegate {
     let response = navigationResponse.response
     let responseURL = response.url
     let tab = tab(for: webView)
+    let isInvalid: Bool
+    if let httpResponse = response as? HTTPURLResponse {
+      isInvalid = httpResponse.statusCode >= 400
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift:604]: Are you sure about this >= 400 check? What happens if the response is a 000?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

**Comment by Brandon-T** [comment]
Needs a review for secure icon state when upgrading from http to https.

File: ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift (hunk 1 above)
Comment (cuba) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/BrowserViewController/BVC+WKNavigationDelegate.swift:604]: I think so: 
https://github.com/brave/brave-core/blob/c4daecc6b66e088f1eee0a3c6bfe04bd0bbba9a5/chromium_src/chrome/browser/ssl/https_upgrades_interceptor.cc#L44

**Review by Brandon-T — APPROVED** [review]

File: ios/browser/api/https_upgrade_exceptions/https_upgrade_exceptions_service.mm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    return nil;
+  }
+
+  // Check url validity
+  if (!gurl.SchemeIs("http") || !gurl.is_valid()) {
+    return nil;
+  }
+
+  // Ensure we didn't add an exception for this domain
+  if (base::Contains(exceptional_domains_, gurl.host())) {
```
Comment (stoletheminerals) [inline ios/browser/api/https_upgrade_exceptions/https_upgrade_exceptions_service.mm:50]: How long does the exception list live?

Comment (cuba) [reply ios/browser/api/https_upgrade_exceptions/https_upgrade_exceptions_service.mm:50]: until you kill the app

**Pull request merged**